primitives/src/merkle_tree/internal.rs

```diff
@@ -327,15 +327,17 @@ where
     }
 }
 
-pub(crate) fn digest_branch<E, H, I, T>(data: &[Box<MerkleNode<E, I, T>>]) -> T
+pub(crate) fn digest_branch<E, H, I, T>(
+    data: &[Box<MerkleNode<E, I, T>>],
+) -> Result<T, PrimitivesError>
 where
     E: Element,
     H: DigestAlgorithm<E, I, T>,
     I: Index,
     T: NodeValue,
 {
     // Question(Chengyu): any more efficient implementation?
-    let data = data.iter().map(|node| node.value()).collect_vec();
+    let data = data.iter().map(|node| node.value()).collect::<Vec<_>>();
     H::digest(&data)
```

Comment:
[out of scope for this PR] Regarding the question: we could change the `DigestAlgorithm` trait so that `digest` takes an `IntoIterator` instead of `&[T]`. Then you could just pass `data.iter().map(|node| node.value())` directly to `digest` with no mem copy or heap alloc.

Reply:
Good idea. Let me do it

Reply:
A tiny issue: `RescueCRHF::sponge_no_padding` is taking a slice of field elements `&[F]`. We still need some memcopy/heap alloc right now. Of course we can change that API in the future.

Do you think it's still worth changing it now?

Reply:
Up to you. It's not high priority. If you don't feel like doing it now then please post a github issue and tag it "low priority" so we don't forget about this.

I have a similar issue with 

https://github.com/EspressoSystems/jellyfish/blob/6641b3ff97bf12d764d7384bc632cdc84c3f8209/utilities/src/conversion.rs#L128

It should take `IntoIterator` as input and return an iterator as output.

I guess we should both post an issue. 😉

Reply:
I think perhaps we should move toward a design pattern of taking `IntoIterator` as input and returning an iterator as output whenever possible. @alxiong any opinion?